### PR TITLE
Add notification messages framework

### PR DESCRIPTION
**Pull request by inevity**
Now we  only impl the NotifyInvalInode,NotifyInvalEntry,
NotifyDelete.

Notify funcs allow the filesystem to invalidate VFS caches.

The statement that 'In a networked fuse file-system, the local
fuse instance may receive changes from a remote system that do not
pass through the local kernel vfs.
When those changes are made the local fuse instance must notify the
local kernel of the change so as to keep the local kernel's cache
in sync. This is accomplished via notify delete or notify
invalidate.
In other words, these notifications only apply if fuse file-system
makes changes to the file-system that are not initiated by the
local kernel: changes initiated by the local kernel must not
involve notify delete or notify invalidate (will deadlock), and
are not required.' is quoted from Jhon Muir.

About fuse  notify func history ,please ref link
http://lkml.iu.edu/hypermail/linux/kernel/0906.0/01120.html

### CODE REVIEW

File: connection.go [hunk 1]
```diff
@@ -241,7 +241,7 @@ func (c *Connection) beginOp(
 	// should not record any state keyed on their ID.
 	//
 	// Cf. https://github.com/osxfuse/osxfuse/issues/208
-	if opCode != fusekernel.OpForget {
+	if opCode != fusekernel.OpForget && opCode < 100 {
```
Comment (jacobsa) [inline connection.go]: Why this change?

File: connection.go [hunk 2]
```diff
@@ -411,6 +411,36 @@ func (c *Connection) ReadOp() (ctx context.Context, op interface{}, err error) {
 	}
 }
 
+func (c *Connection) SetNotifyContext(op interface{}) (context.Context, error) {
+
```
Comment (jacobsa) [inline connection.go]: Please remove the blank line here. (Please run gofmt if you haven't.)

File: connection.go [hunk 3]
```diff
@@ -411,6 +411,36 @@ func (c *Connection) ReadOp() (ctx context.Context, op interface{}, err error) {
 	}
 }
 
+func (c *Connection) SetNotifyContext(op interface{}) (context.Context, error) {
```
Comment (jacobsa) [inline connection.go]: Please add documentation for all new exported symbols.

File: fuseops/ops.go [hunk 4]
```diff
@@ -845,3 +845,24 @@ type SetXattrOp struct {
 	// simply replace the value if the attribute exists.
 	Flags uint32
 }
+type NotifyPollOp struct {
```
Comment (jacobsa) [inline fuseops/ops.go]: It's especially important to have good documentation for the new Op structs. See the other ops in this package for an example of the kind of thing I'm looking for.

**Comment by inevity** [comment]
Have added documentation.

File: connection.go (hunk 1 above)
Comment (inevity) [reply connection.go]: For notify ops ,it is issued by the userspace filesystem to
fuse kernel. The ops's kernel opCode is 0 in kernel' viewpoint or kernel not
care whether what the opcode is .But the ops have the notifycode  from 1
to 6 now. For adapt the notfiy ops to the BeginOP ,we fake an opcode as
notfiycode + 100. Now opcode range of ops from kernel including origin
 from userspace and just from kernel is < 100. The notify ops's opcode
range is [101-106].
Then we can process all ops consistently.
Cf. func (c *Connection) SetNotifyContext(op interface{})
            (context.Context, error) {

see pr code.

File: connection.go (hunk 3 above)
Comment (inevity) [reply connection.go]: Done.

File: fuseops/ops.go (hunk 4 above)
Comment (inevity) [reply fuseops/ops.go]: Done.

**Comment by fishpadak** [comment]
Hello. Any progress on this PR?

**Comment by inevity** [comment]
> Hello :D Any progress on this PR?

Dont know why jacobsa not to merge or give some advice.

**Comment by stapelberg** [comment]
Would you mind rebasing this on the current master branch please? We fixed the travis config in the meantime.

**Comment by inevity** [comment]
Just have rebased on master branch.

File: connection.go [hunk 5]
```diff
@@ -497,6 +541,29 @@ func (c *Connection) Reply(ctx context.Context, opErr error) {
 	}
 }
 
+// The NotifyKernel is same as Reply func of Connection.But the diff is
+// that the func only send to kernel.
+func (c *Connection) NotifyKernel(ctx context.Context) {
+
+	// we should get outmsg from context
+	var key interface{} = contextKey
+	foo := ctx.Value(key)
```
Comment (stapelberg) [inline connection.go]: Can you explain why we need to pull this value out of the context instead of just passing it as a parameter to NotifyKernel?

Comment (inevity) [reply connection.go]: Just do like the reply func.  Of course you can also just pass the value of the context as a parameter to NotifyKernel.

Comment (stapelberg) [reply connection.go]: Thanks. The reply mechanism needs to use a context because it passes a value (https://github.com/jacobsa/fuse/blob/659cc51b20ee0bf7bf68875df8ea1a0e0704ba42/connection.go#L406) across API boundaries (to the file system implementation and back).

This is not the case here AIUI, so please change the function signature to not take a context, but just take the opState parameter directly.

Comment (inevity) [reply connection.go]: I rethink this case, even the notify func is  diff with the normal op which need reply function. But the context use in notify func  is same as the reply func.   Just the context is created by our func (s *fileSystemServer) Invalidateentry/inode.   And now we call unity all notify op and the normal op.  
For the correct context use, we should alway use the context for notify case. see func (s *fileSystemServer) InvalidateInode for opsInFlight.Done.  

Maybe i am not understandering the issue on detail.

Comment (stapelberg) [reply connection.go]: > I rethink this case, even the notify func is diff with the normal op which need reply function. But the context use in notify func is same as the reply func. Just the context is created by our func (s *fileSystemServer) Invalidateentry/inode. And now we call unity all notify op and the normal op.

If I understand correctly, you’re saying having both functions (NotifyKernel and Reply) take arguments via a context is good because they are similar, yes?

While similarity is good, I would argue that in this case, it’s clearer if a context is not used unless absolutely necessary. Since we’re creating a context but then immediately call the function (instead of passing the context somewhere else), I don’t think we need to use a context.

Contexts are used (aside from deadlines and cancellation) to propagate data across API boundaries. In this case, we don’t need to cross API boundaries, so we shouldn’t use a context if we can use a simple variable instead.

**Comment by inevity** [comment]
Can you give some comments about the context commit?
I use context in SetNotifyContext,so this func SetNotifyContext  have nothing to modify, only one bug to fix.

Then change NotifyKernel(opstate opState) as this signature.

And the funcs which  calling notifykernel  in fuseutil/file_system.go Where have some dup code, you can give some detail advice.

**Comment by inevity** [comment]
Wait...

File: fuseutil/file_system.go [hunk 6]
```diff
@@ -123,6 +125,77 @@ func (s *fileSystemServer) ServeOps(c *fuse.Connection) {
...
+	c := s.GetMfs().Conn.(*fuse.Connection)
+	op := &fuseops.NotifyInvalInodeOp{
+		Ino: fuseops.InodeID(ino),
+		Off: off,
+		Len: len,
+	}
+	ctx, _ := c.SetNotifyContext(op)
+	var key interface{} = contextKey
+        foo := ctx.Value(key)
+        opstate, ok := foo.(opState)
```
Comment (riking) [inline fuseutil/file_system.go:187]: The formatting here is a mess.

File: mount.go [hunk 7]
```diff
@@ -27,6 +28,11 @@ type Server interface {
 	// until all operations have been responded to. Must not be called more than
 	// once.
 	ServeOps(*Connection)
+	InvalidateEntry(fuseops.InodeID, string) error
+	InvalidateInode(fuseops.InodeID, int64, int64) error
+	NotifyDelete(fuseops.InodeID, fuseops.InodeID, string) error
+	SetMfs(*MountedFileSystem)
+	GetMfs() *MountedFileSystem
 }
```
Comment (riking) [inline mount.go:36]: This is a lot of extra methods to be adding to the very simple Server interface, which could plausibly have been proxied in client code..
